### DISCUSSION TITLE
Issue with embed_migrations! macro resolving relative path during cargo publish

### DISCUSSION THREAD

**Discussion by martsokha** [diesel-rs/diesel#4389] [Q&A] [answered]
Hello,

I'm encountering an issue when using the `embed_migrations!` macro in my project. The code works fine during `cargo clippy`, but when I attempt to publish the crate with `cargo publish`, I receive the following error:

```rust
pub const MIGRATIONS: EmbeddedMigrations = embed_migrations!("../../migrations");
```

```
error: proc macro panicked
 --> config/embedded_query.rs:6:44
  |
6 | pub const MIGRATIONS: EmbeddedMigrations = embed_migrations!("../../migrations");
  |                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: message: Failed to receive migrations dir from Some("../../migrations")
```

My workspace structure looks like this:

```
./Cargo.toml
./Cargo.lock
./crates
  ./crates/schema
    ./crates/schema/Cargo.toml
    ./crates/schema/Cargo.lock
  ./crates/migrate
    ./crates/migrate/Cargo.toml
    ./crates/migrate/Cargo.lock
./migrations
```

How can I ensure that the `embed_migrations!` macro correctly resolves the migrations directory path when publishing the crate?

Thanks in advance for any help or suggestions!

**Comment by weiznich** [comment, marked as answer]
I can reproduce this, but I fear there is not much we as diesel team can do to resolve this. It seem more like an issue with your code structure.

The underlying problem is that if you provide an explicit relative path to the migrations directory like that the proc macro will resolve that directory starting from the crate root. Now cargo publish copies everything (according to the include/exclude rules in your `Cargo.toml` file) in your crate root (everything in `crates/migrate` in the case of the migrate crate) into the `target/package` and tries to build it from there. That then fails as the relative path to the migration directory is now not correct anymore as the migration directory is **outside** the crate root and wouldn't be included into the published crate anyway.

**Reply by martsokha** [reply]
Thanks for the detailed response. I ended up resolving the issue by nesting the migrations folder, but it does feel a bit wrong.

I was honestly hoping for a more flexible solution: something similar to what `tonic` does, where the migrations folder and its contents could be copied to the `OUT_DIR` in a `build.rs` script and then passed to `diesel_migrations`. 

Unfortunately, it doesn’t seem like there’s a way to pass environment variables to the `embed_migrations!` macro.

**Reply by weiznich** [reply]
I don't think a more flexible solution would solve the underlying problem. It might help passing the build done by `cargo publish` but you would still end up with a broken package. `cargo publish` would still only include files in the same directory as your `Cargo.toml` file, which would exclude the `migrations` directory. That means anyone that would try to use your crate would get a build error due to the missing migration files. There is unfortunately not much we can do in diesel to address this problem as there is no way for us to instruct cargo to also include these migration files.
